torch/_inductor/triton_bundler.py

```diff
@@ -0,0 +1,234 @@
+    their artifacts into a structured bytes blob, and later when write is called
+    it writes this structured blob back to file system.
+
+    Intended Life cycle:
+    - TritonBundler.begin_compile is called when we start compiling in Inductor
+    - TritonBundler.put is called each time a Triton Kernel is compiled
+    - TritonBundler.collect is called when a cache entry is being generated
+    - TritonBundler.read_and_emit is called when a cache entry is read
+    """
+
+    _entries: Optional[List[TritonBundleEntry]] = None
+
+    # __grp__kernel_name.json contains metadata with source code paths
+    # we use this as sentinal value for search and replace
+    _REPLACE_BYTES: bytes = b"[REPLACE]"
+
+    @staticmethod
+    def is_enabled() -> bool:
+        from torch._inductor import config
+
+        if config.force_disable_caches:
+            return False
+
+        if (b := config.bundle_triton_into_fx_graph_cache) is not None:
+            return b
+
+        if not config.is_fbcode():
+            return False
+
+        return justknobs_check("pytorch/remote_cache:bundle_triton_into_fx_graph_cache")
+
+    @classmethod
+    def begin_compile(cls) -> None:
+        """
+        Initializes the TritonBundler.
+        The current TritonBundler bundle is finalized by TritonBundler.collect.
+        """
+        if not TritonBundler.is_enabled():
+            return
+        assert cls._entries is None
+        cls._entries = []
+
+    @classmethod
+    def put(cls, kernel_hash: str, device: int) -> None:
+        """
+        Lazily observes that we have seen a Triton kernel compilation. Remembers
+        it for when collect is later called.
+        """
+        if (entries := cls._entries) is not None:
+            entries.append(
+                TritonBundleEntry(kernel_hash, device, triton_cache_dir(device))
+            )
+
+    @classmethod
+    def collect(
+        cls,
+    ) -> Tuple[List[TritonKernelArtifacts], Optional[TritonBundlerMetadata]]:
+        """
+        This is the main function called when a cache write happens. This function
+        converts all the previously remembered kernels into bundled format so that
+        it can be written into a cache entry.
+        This function also finalizes the current bundle.
+        """
+        if not TritonBundler.is_enabled():
+            cls._entries = None
+            return [], None
+
+        with dynamo_timed(key="TritonBundler.collect", fwd_only=False):
+            entries = cls._entries
+            if entries is not None:
+                result: List[TritonKernelArtifacts] = []
+                kernel_names: List[str] = []
+                for entry in entries:
+                    artifacts: List[TritonKernelArtifact] = []
+                    path = os.path.join(entry.directory, entry.kernel_hash)
+                    if not os.path.exists(path):
+                        continue
+                    for filename in os.listdir(path):
```

Comment:
So post our discussion, we have exclusive access to the kernel directory at this point in time (no one potentially racing us writing files into this directory while we are collecting).

Reply:
i added this to the doc block above